### PR TITLE
 compile and install the static library of fluid inference

### PR DESCRIPTION
**Pull request by luotao1**
related #7231
- The structure of the installed shared library is as follows. It uses default cmake option + `-DWITH_PYTHON=OFF`
```
PaddleRoot/
├── bin
│   └── paddle
├── include
│   └── paddle
│       ├── framework
│       ├── inference
│       ├── memory
│       ├── platform
│       └── string
├── lib
│   ├── libpaddle_fluid.a
│   └── libpaddle_fluid.so
├── opt
│   └── paddle
│       └── bin
└── third_party
    ├── eigen3
    │   ├── Eigen
    │   └── unsupported
    ├── gflags
    │   ├── include
    │   └── lib
    ├── glog
    │   ├── include
    │   └── lib
    ├── openblas
    │   ├── include
    │   └── lib
    └── protobuf
        ├── include
        └── lib
```
```
du -sh *
12K     bin
700K    include
188M    lib
200M    opt
92M     third_party
```
```
128M    libpaddle_fluid.a
61M     libpaddle_fluid.so
```
- The DEPS of static library is manually added one by one now, it always missing some new DEPS. So adding DEPS automatically will be done later. 
- Examples: https://github.com/luotao1/fluid_inference_example.

### CODE REVIEW

**Comment by luotao1** [comment]
> Please refer to my comment #7231 (comment) 8 days ago

Discussed with @Xreki, we have already answered in https://github.com/PaddlePaddle/Paddle/pull/7572#discussion_r161982735, but sorry to forget @ you.

> It seems that this PR and the issue aforementioned are related to the binary release of PaddlePaddle.

The binary directory `/bin` and `/opt` is left for C-API since `WITH_FLUID` option (#6346) has not been finished now. After finishing #6346, the final fluid inference library directory likes as follows, and it is not a binary library:
```
PaddleRoot/
├── include
│   └── paddle
│       ├── framework
│       ├── inference
│       ├── memory
│       ├── platform
│       └── string
├── lib
│   ├── libpaddle_fluid.a
│   └── libpaddle_fluid.so
└── third_party
    ├── eigen3
    │   ├── Eigen
    │   └── unsupported
    ├── gflags
    │   ├── include
    │   └── lib
    ├── glog
    │   ├── include
    │   └── lib
    ├── openblass
    │   ├── include
    │   └── lib
    └── protobuf
        ├── include
        └── lib
```
> However, to what I understand is that our binary release would be in the form of pip packages

As both IOS and Android don't support pip packages since they don't support python, the fluid inference library should not be in the form of this way.

> Also, even if we are going to add binary libraries as an additional form of binary release, we are not supposed to put third-party binaries in our directory hierarchy.

Our third-party libraries only have `include/` and `lib/`, and don't contain any binaries.

There are two reasons why we put third-party libraries in our directory:
- For some platforms (such as IOS) and usages, users must link static library. Thus, they must link the corresponding third-party libraries. If we don't provide them, users must download and compile them manually, which is not convenient and sometimes hard for users. For example, like `openblas.cmake`, the compile option is different from IOS, Android, RPI, Apple and Linux.
https://github.com/PaddlePaddle/Paddle/blob/b4685675df63d7331e8215c36fff66eacde5da7e/cmake/external/openblas.cmake#L35-L67
- It is easy for users to know how many third-party libraries depended by paddle, and it is easy for users to link these libraries themselves.

> The right way is to build only ours and pack a .deb file, which describes on what other .deb files it depends.

`.deb` can only be installed on Ubuntu, but it can't be installed on Mac, IOS and Andriod. Thus, `.deb` may not be a good choice.

**Comment by wangkuiyi** [comment]
I agree that we publish our inference library, but with two points different from the current status of this PR:

1. `make inference_lib_dist` instead of `make install`
1. the directory hierarchy in the generated tgz file should be the same as that in `./build`.

**Comment by luotao1** [comment]
@wangkuiyi 
> the directory hierarchy in the generated tgz file should be the same as that in ./build

If the same as that in ./build, the final directory hierarchy would like as follows:

```
PaddleRoot
├── paddle
│   ├── framework
│   │   ├── attribute.h
│   │   ├── backward.h ...
│   │   └── var_type_inference.h
│   ├── inference
│   │   ├── inference.h
│   │   ├── libpaddle_fluid.a
│   │   └── libpaddle_fluid.so
│   ├── memory
│   │   ├── detail
│   │   ├── memcpy.h
│   │   └── memory.h
│   ├── platform
│   │   ├── assert.h
│   │   ├── call_once.h ...
│   │   └── variant.h
│   └── string
│       ├── piece.h
│       ├── printf.h
│       ├── tinyformat
│       └── to_string.h
└── third_party
    ├── eigen3
    │   ├── Eigen
    │   └── unsupported
    ├── gflags
    │   ├── include
    │   └── lib
    ├── glog
    │   ├── include
    │   └── lib
    ├── openblas
    │   ├── include
    │   └── lib
    └── protobuf
        ├── include
        └── lib
```

File: cmake/external/openblas.cmake [hunk 1]
```diff
@@ -82,6 +82,7 @@ IF(NOT ${CBLAS_FOUND})
         CONFIGURE_COMMAND   ""
     )
     SET(CBLAS_PROVIDER openblas)
+    FILE(REMOVE_RECURSE ${CBLAS_INSTALL_DIR}/lib/cmake ${CBLAS_INSTALL_DIR}/lib/pkgconfig)
```
Comment (Xreki) [inline cmake/external/openblas.cmake]: `FILE`命令应该是在执行`cmake`命令的时候执行的。`extern_openblas`是在`make`阶段执行的，也就是说，必须要执行完`extern_openblas`这个任务之后，才会生成`${CBLAS_INSTALL_DIR}/lib/cmake`和`${CBLAS_INSTALL_DIR}/lib/pkgconfig`这两个目录。所以这条语句应该达不到删除的目的。应该使用`execute_process`语句删除，并且该语句需要依赖`extern_openblas`这个任务。

File: paddle/inference/CMakeLists.txt [hunk 2]
```diff
@@ -4,8 +4,20 @@ cc_library(paddle_fluid_api
     SRCS io.cc
     DEPS ${FLUID_CORE_MODULES} ${GLOB_OP_LIB})
 
+if(WITH_GPU)
+  set(GPU_DEPS gpu_info dynload_cuda)
+endif()
 # Merge all modules into a single static library
-cc_library(paddle_fluid DEPS paddle_fluid_api ${FLUID_CORE_MODULES} ${GLOB_OP_LIB})
+cc_library(paddle_fluid DEPS paddle_fluid_api
```
Comment (Xreki) [inline paddle/inference/CMakeLists.txt]: 像`memory`模块一样，我们在每个模块最后加一个`cc_library`，将该模块里面所有的target都合并成一个静态库？

https://github.com/PaddlePaddle/Paddle/blob/1185a1b5ab897133585ffdcef8f68d0ee6caef46/paddle/memory/CMakeLists.txt#L6-L14

这样，在inference里面，我们只需要依赖`paddle_memroy`, `paddle_framework`, `paddle_operators`, `paddle_string`, `paddle_platform`。不过这样还是不能自动添加。

File: cmake/inference_lib.cmake [hunk 3]
```diff
@@ -45,11 +58,19 @@ copy(glog_lib
 IF(NOT PROTOBUF_FOUND)
     set(dst_dir "${CMAKE_INSTALL_PREFIX}/third_party/install/protobuf")
     copy(protobuf_lib
-      SRCS ${PROTOBUF_INCLUDE_DIR} ${PROTOBUF_LITE_LIBRARY}
+      SRCS ${PROTOBUF_INCLUDE_DIR} ${PROTOBUF_LIBRARY}
       DSTS ${dst_dir} ${dst_dir}/lib
     )
 ENDIF(NOT PROTOBUF_FOUND)
 
+IF(NOT WITH_MKL)
```
Comment (Xreki) [inline cmake/inference_lib.cmake]: 这里应该是`if(NOT CBLAS_FOUND)`，`cblas.cmake`里面会去找系统里的一些blas库，只有`CBLAS_FOUND`为`OFF`时才会自动去下载、编译和安装openblas。原则上，不是Paddle自动下载、编译的库，我们不要拷贝吧。`mklml`在加入的时候，也在`cblas.cmake`里面设置了`CBLAS_FOUND`，这种情况另外再考虑吧。

另外，大小写统一一下？:smile:

Comment (luotao1) [reply cmake/inference_lib.cmake]: Done

**Pull request merged**